### PR TITLE
Hookup create sections FE/BE

### LINKED ISSUE #88: Integrate the FE with BE for create sections
Sections creation (~60) takes about 2 min even if all failed and retry it takes about ~5sec. So user gets an immediate response about the status of the request.

1. Once section creating is finished, the BE is sending what sections failed and what is the reasons for failure. The FE needs to display the section/reason for failure in a tabular form. This was something discussed with Janel/Melinda
2. When GetSection call failed, the user needs to be presented with the reason for failure of that section that is coming from the BE. "Failed to get the exiting section, please try again later" He should not be taken to the creation section upload screen. 
3. For all success case we tell user that with a generic message that "All sections are successfully created". not need to present with all the list of section that was successful

Response object from Canvas
```
{
  "statusCode": 422,
  "errors": [
    {
      "failedInput": "sec1",
      "message": "Error(s) from Canvas:  Unprocessable Entity"
      "canvasStatusCode": 422
    }
    
  ]
}
```
4. NestJs does some basic validation to check if the string is not empty and should not exceed 250 in length. Below is the Response structure from NestJS error. Some thought needs to be given in `handleErrors` how to handle different error object or FE probably could also do the validation for empty and length > 250. Doing both would be redundant. 
Response object from NestJs
```
{
  "statusCode": 400,
  "message": [
    "each value in sections must be shorter than or equal to 250 characters",
    "each value in sections should not be empty"
  ],
  "error": "Bad Request"
}
```

### PR DESCRIPTION
**Pull request by chrisrrowland**
Closes #88

### CODE REVIEW

File: ccm_web/client/src/api.ts [hunk 1]
```diff
@@ -55,18 +65,23 @@ export const getGlobals = async (key: string | undefined): Promise<Globals> => {
...
-      return new Promise<string[]>((resolve, reject) => { reject(new Error('Error retrieving course section information.')) })
-    }
-  })
-  return sections
+export const addCourseSections = async (key: string | undefined, courseId: number, sectionNames: string[]): Promise<CanvasCourseSection[]> => {
+  const body = JSON.stringify({ sections: sectionNames })
+  const request = getPost(key, body)
+  const resp = await fetch('/api/course/' + courseId.toString() + '/sections', request)
+  await handleErrors(resp)
+  return await resp.json()
```
Comment (pushyamig) [inline ccm_web/client/src/api.ts:86]: I see this error in the browser console when getting sections are loading. Probably I feel this is due the way the promise is handled
```
react-dom.development.js:67 Warning: findDOMNode is deprecated in StrictMode. findDOMNode was passed an instance of Transition which is inside StrictMode. Instead, add a ref directly to the element you want to reference. Learn more about using refs safely here: https://reactjs.org/link/strict-mode-find-node
    at div
    at Transition (https://925849cc5a54.ngrok.io/main-00f62ec4d542a7626326.js:57029:30)
    at Fade (https://925849cc5a54.ngrok.io/main-00f62ec4d542a7626326.js:3573:24)
    at Backdrop (https://925849cc5a54.ngrok.io/main-00f62ec4d542a7626326.js:632:24)
    at WithStyles (https://925849cc5a54.ngrok.io/main-00f62ec4d542a7626326.js:19027:31)
    at div
    at div
    at div
    at BulkSectionCreate (https://925849cc5a54.ngrok.io/main-00f62ec4d542a7626326.js:64493:21)
```

File: ccm_web/client/src/pages/BulkSectionCreate.tsx [hunk 2]
```diff
@@ -415,18 +506,42 @@ Section 001`
...
 
+  const renderSuccess = (): JSX.Element => {
+    return (
+      <Card className={successClasses.card} variant="outlined">
+        <CardContent className={successClasses.cardContent}>
+          <CheckCircle className={successClasses.icon} fontSize='large'/>
+          <Typography>New sections have been added!</Typography>
+        </CardContent>
+        <CardActions className={successClasses.cardFooter}>
+          See your sections in the Canvas Setting
```
Comment (pushyamig) [inline ccm_web/client/src/pages/BulkSectionCreate.tsx:540]: We need to provide a link to the page setting `/courses/:course_id/settings#tab-sections` to see if sections has been added or not.

Comment (chrisrrowland) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:540]: This requires additional functionality that were going to handle in a separate issue.  The canvas url needs to be loaded into globals

File: ccm_web/client/src/api.ts (hunk 1 above)
Comment (chrisrrowland) [reply ccm_web/client/src/api.ts:86]: I think this is from a 3rd party library but I can check again

**Comment by pushyamig** [comment]
There are some alignment issue with the way file name is displayed

Screen Shot 2021-07-28 at 8 23 38 AM

File: ccm_web/client/src/pages/BulkSectionCreate.tsx (hunk 2 above)
Comment (pushyamig) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:540]: The globals already is sending the `course_id`, why can't we make the URL like `/courses/403334/setting#tab-sections`?

global call response
`{"environment":"development","userLoginId":"dddd","course":{"id":403334,"roles":["TeacherEnrollment","Account Admin"]}}`

File: ccm_web/client/src/api.ts [hunk 3]
```diff
@@ -55,18 +65,23 @@ export const getGlobals = async (key: string | undefined): Promise<Globals> => {
   return await resp.json()
 }
 
-const delay = async (ms: number): Promise<void> => {
-  await new Promise<void>(resolve => setTimeout(() => resolve(), ms))
+// usage:
+//   const data = await delay(nnnn).then(() => {return something})
+// const delay = async (ms: number): Promise<void> => {
+//   await new Promise<void>(resolve => setTimeout(() => resolve(), ms))
+// }
```
Comment (pushyamig) [inline ccm_web/client/src/api.ts:72]:
```suggestion

```

File: ccm_web/client/src/models/canvas.ts [hunk 4]
```diff
@@ -2,3 +2,9 @@ export interface CanvasCourseBase {
   id: number
   name: string
 }
+
+export interface CanvasCourseSection {
+  id: number
+  name: string
+  // total_students?: number
```
Comment (pushyamig) [inline ccm_web/client/src/models/canvas.ts:9]: why is this commented out? This should be included

File: ccm_web/client/src/pages/BulkSectionCreate.tsx (hunk 2 above)
Comment (chrisrrowland) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:540]: We know the ending of the url but not the environment dependent beginning

File: ccm_web/client/src/models/canvas.ts (hunk 4 above)
Comment (chrisrrowland) [reply ccm_web/client/src/models/canvas.ts:9]: It was not used in my changes so I left it to be added later...

**Comment by pushyamig** [comment]
I found this interesting case when validation error is caught saying sections already occurred by so and so name in course, I opened another window i deleted sections and got back to same CCM window with validation error for create section and tried to upload again but CCM state still thinks the sections loaded already exits. How should we handle this case? which to me not usually a regular workflow that subaccount admins will do ( like delete the sections)

**Comment by chrisrrowland** [comment]
> I found this interesting case when validation error is caught saying sections already occurred by so and so name in course, I opened another window i deleted sections and got back to same CCM window with validation error for create section and tried to upload again but CCM state still thinks the sections loaded already exits. How should we handle this case? which to me not usually a regular workflow that subaccount admins will do ( like delete the sections)

Clicking the fix and upload again button could refresh the existing course list.

Probably some other options too

File: ccm_web/client/src/api.ts (hunk 3 above)
Comment (chrisrrowland) [reply ccm_web/client/src/api.ts:72]: Are code comments considered poor practice now?  I feel like any time I leave commented code it's asked to be removed.

**Comment by pushyamig** [comment]
>I found this interesting case when validation error is caught saying sections already occurred by so and so name in course, I opened another window i deleted sections and got back to same CCM window with validation error for create section and tried to upload again but CCM state still thinks the sections loaded already exits. How should we handle this case? which to me not usually a regular workflow that subaccount admins will do ( like delete the sections)
Clicking the fix and upload again button could refresh the existing course list.

Probably clicking the upload button could get the section again? could be easiest thing to do

File: ccm_web/client/src/api.ts (hunk 3 above)
Comment (pushyamig) [reply ccm_web/client/src/api.ts:72]: Generally I won't leave code that is not used in the repo, But you can ask other developers what is the best practice here

File: ccm_web/client/src/pages/BulkSectionCreate.tsx (hunk 2 above)
Comment (pushyamig) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:541]: relative URL's won't work in this case? just like `<a href = '/course/40404/setting#tab-section'` or using a link Material UI component?

**Comment by pushyamig** [comment]
@chrisrrowland you can work with other team member for getting approval since I am OOO next week.

File: ccm_web/client/src/api.ts (hunk 1 above)
Comment (chrisrrowland) [reply ccm_web/client/src/api.ts:86]: I found what appears to be a solution for it, though I think it is simply just conditionally using a workaround in development.  I'm not sure if it makes it go away in production or not.

However, it is an issue with a 3rd party library and material-ui so it doesn't really belong in this issue.

The workaround involves using the  ThemeProvider so it might be best handled as part of issue #75  for implementing an application level theme.

I moved the SnackbarProvider up to Index.tsx from App.tsx and wrapped it in ThemeProvider.
This solution is talked about [here](https://github.com/mui-org/material-ui/issues/13394#issuecomment-815452717) on a material-ui issue.

```
import { createMuiTheme, ThemeProvider, unstable_createMuiStrictModeTheme } from '@material-ui/core'

const ltiKey = getLTIKey()

const createTheme = process.env.NODE_ENV === 'production' ? createMuiTheme : unstable_createMuiStrictModeTheme
const theme = createTheme({

})

ReactDOM.render(
  <React.StrictMode>
    <ThemeProvider theme={theme}>
      <SnackbarProvider maxSnack={3}>
        <App ltiKey={ltiKey} />
      </SnackbarProvider>
    </ThemeProvider>
  </React.StrictMode>,
  document.getElementById('root')
)
```

**Review by lsloan — COMMENTED** [review]
I don't have much time to complete this review at this time, but I'll make more time this afternoon.  In the meantime, maybe @ssciolla will review and approve.  I just have a couple of comments to make at the moment, because I need to take my dog to the vet.

**Review by ssciolla — COMMENTED** [review]
Looking good, @chrisrrowland. Lots of small comments. I think there are a couple areas worth addressing -- related to reducing duplication in validation removing unused stuff, fixing important text, or streamlining a `useEffect`. I'm going to test and open that settings URL PR now, and I'll approve if I'm satisfied with how it works. Feel free to discuss any of the below with me in the meantime.

File: ccm_web/client/src/pages/BulkSectionCreate.tsx [hunk 5]
```diff
@@ -415,32 +536,57 @@ Section 001`
                 <Typography>Review your CSV file</Typography>
                 <CloudDoneIcon className={confirmationClasses.dialogIcon} fontSize='large'/>
                 <Typography>Your file is valid!  If this looks correct proceed with download</Typography>
```
Comment (ssciolla) [inline ccm_web/client/src/pages/BulkSectionCreate.tsx:538]:
```suggestion
                <Typography>Your file is valid!  If this looks correct, click "Submit" to proceed.</Typography>
```

**Review by ssciolla — COMMENTED** [review]
I tested and submitted a PR for the settings URL (chrisrrowland#3). I flagged a couple more things around the validation flow and error handling that I'd like to discuss. Generally the code works well and handles most of the cases fluently. I will wait til I've at least heard your opinion about some of my comments, then I'll approve. I hope the detailed review feels at least somewhat productive to you here.

File: ccm_web/client/src/api.ts [hunk 6]
```diff
@@ -23,6 +23,16 @@ const getGet = (key: string | undefined): RequestInit => {
...
+const getPost = (key: string | undefined, body: string): RequestInit => {
+  const headers: string[][] = []
+  headers.push(['Content-Type', 'application/json'])
+  headers.push(['Accept', 'application/json'])
+  const request = initRequest(key, headers)
+  request.method = 'POST'
+  request.body = body
+  return request
+}
+
```
[the remark this answers is not in the document]
Comment (lsloan) [reply ccm_web/client/src/api.ts:35]: That's a good suggestion.  My IDE complains that some of the code I'm using for section enrollment dupes the section creation.  I was thinking of seeing whether that could be combined.

File: ccm_web/client/src/api.ts (hunk 3 above)
Comment (chrisrrowland) [reply ccm_web/client/src/api.ts:72]: Don't they just get minified out in production anyways?

I might remember that once upon a time I left myself a comment somewhere (Narrator: he won't) and look it up in the history, but nobody else is going to know that once upon a time there was a comment that might have been helpful when they are looking at some code for the first time.

tldr; get off my lawn ;-)

File: ccm_web/client/src/models/canvas.ts (hunk 4 above)
Comment (chrisrrowland) [reply ccm_web/client/src/models/canvas.ts:9]: Sometimes I put in code that isn't used yet and I get asked to take it out, so I thought this was a happy medium, but I'll just uncomment it.

File: ccm_web/client/src/pages/BulkSectionCreate.tsx [hunk 7]
```diff
@@ -239,27 +329,30 @@ function BulkSectionCreate (props: BulkSectionCreateProps): JSX.Element {
...
+    return rowInvalidations
+  }
+
+  const doServerValidation = (): SectionsRowInvalidation[] => {
+    const rowInvalidations: SectionsRowInvalidation[] = []
+    const duplicatesNamesInCanvasValidator: DuplicateExistingSectionRowsValidator = new DuplicateExistingSectionRowsValidator()
+    rowInvalidations.push(...duplicatesNamesInCanvasValidator.validate(sectionNames))
+
+    return rowInvalidations
+  }
```
[the remark this answers is not in the document]
Comment (chrisrrowland) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:352]: One validates against sections that exist in the local file, one validates against duplicates that exist in  canvas.  They could be designed to use a single validator by making the validator itself more complicated but I chose to do it this way

[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:352]: Ah, I'll be honest, I didn't look too closely at the exact implementations until this morning. Some similarities, but I take your point that it would make it more complicated. I'm okay keeping them separate.

That said, I see you are using `existingSectionNames` from the component scope directly in the validator. Could we make that an additional method parameter for `validate` (or a constructor parameter, instance variable) and then move this validator to be with the others in `BulkSectionCreateValidators.ts`?

File: ccm_web/client/src/pages/BulkSectionCreate.tsx [hunk 8]
```diff
@@ -147,24 +184,66 @@ function BulkSectionCreate (props: BulkSectionCreateProps): JSX.Element {
...
   useEffect(() => {
+    if (pageState.state === BulkSectionCreatePageState.Saving) {
+      const serverInvalidations = doServerValidation()
+      if (serverInvalidations.length !== 0) {
+        handleRowLevelInvalidationError(serverInvalidations)
+      } else {
+        void doSaveCanvasSectionData()
+      }
+    }
+  }, [existingSectionNames])
```
[the remark this answers is not in the document]
Comment (chrisrrowland) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:214]: The original PR worked that way.  @pushyamig brought up a scenario where the state of canvas would change while the user had the create sections window open, so the recent work was to change the flow to only validate against the current canvas state on submit.

File: ccm_web/client/src/models/models.ts [hunk 9]
```diff
@@ -42,3 +42,7 @@ export interface APIErrorData {
   statusCode: number
   errors: APIErrorPayload[]
 }
+
+export interface IDefaultError {
```
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/models/models.ts:46]: Ah, thought it might be a Java thing.

File: ccm_web/client/src/pages/BulkSectionCreate.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx:214]: Ah, I see. I guess that makes sense. I'll resolve this.

File: ccm_web/client/src/pages/BulkSectionCreate.tsx [hunk 10]
```diff
@@ -415,32 +536,57 @@ Section 001`
                 <Typography>Review your CSV file</Typography>
                 <CloudDoneIcon className={confirmationClasses.dialogIcon} fontSize='large'/>
                 <Typography>Your file is valid!  If this looks correct proceed with download</Typography>
-                <Button variant="outlined" onClick={(e) => resetPageState()}>Do Something</Button>
+                <Button variant="outlined" onClick={(e) => resetPageState()}>Cancel</Button>
+                <Button variant="outlined" onClick={async () => await submit()}>Sumbit</Button>
```
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx]: Looks like you edited something here, but we still should fix the spelling of "Submit" and the arrow function wrapper is redundant, I believe.

File: ccm_web/client/src/models/canvas.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/models/canvas.ts]: Not sure if we ever created an issue for this. I did look into it, hopefully we'll get to it at some point.

**Review by ssciolla — APPROVED** [review]
Cool, thanks for going through my comments, I'm approving. Feel free to handle anything remaining as you see fit. Questions about validation errors from the backend and the settings link can be handled later (or separately). 🚀

File: ccm_web/client/src/pages/BulkSectionCreate.tsx (hunk 10 above)
[the remark this answers is not in the document]
Comment (ssciolla) [reply ccm_web/client/src/pages/BulkSectionCreate.tsx]: Spelling fixed, arrow function still there, but not a big deal.
